File: (19)hasSubTree/Solution.java
/*
树的子结构
输入两颗二叉树A，B，判断B是不是A的子结构。
*/
class TreeNode {

    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;
    }
}

public class Solution {
    public static boolean HasSubtree(TreeNode root1,TreeNode root2) {
    	boolean result = false;
    	if(root1!=null&&root2!=null){
    	if(root1.val==root2.val)
        	result = doesHaveSubTree(root1,root2);
        if(!result)
        	result = HasSubtree(root1.left,root2);
        if(!result)
        	result = HasSubtree(root1.right,root2);
    	}
    	return result;
        }

    public static boolean doesHaveSubTree(TreeNode root1,TreeNode root2){
    	if(root2 == null)
    		return true;
    	if(root1 == null)
    		return false;
    	if(root1.val!= root2.val)
    		return false;
    	return (doesHaveSubTree(root1.left,root2.left)&&doesHaveSubTree(root1.right,root2.right));
    }

        public static void main (String [] args){
        	TreeNode tree1 = new TreeNode(1);
        	TreeNode tree2 = new TreeNode(2);
        	System.out.println(HasSubtree(tree1,tree1));
        }
    }

  /*
  1.涉及到指针的，尤其要注意的是程序的鲁棒性
  2.和指针的当前实际指向
  3.彻底理清思路然后再下手，不要着急动手去写，思路打通
  */

File: (19)hasSubTree/Solution.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
    	if(pRoot2 == NULL || pRoot1 == NULL)
    		return false;
    	if(pRoot1->val == pRoot2->val) {
    		if(HasSubtree2(pRoot1->left, pRoot2->left) && HasSubtree2(pRoot1->right, pRoot2->right)) {
    			return true;
    		}
    	}
    	return HasSubtree(pRoot1->left, pRoot2) || HasSubtree(pRoot1->right, pRoot2);
   }

public:
	bool HasSubtree2(TreeNode* pRoot1, TreeNode* pRoot2)
	{
		if(pRoot2  == NULL) {
    		return true;
    	}

    	if(pRoot1 == NULL) {
    		return false;
    	}
    
    	if(pRoot1->val != pRoot2->val) {
    		return false;
    	} else {
    		return HasSubtree2(pRoot1->left, pRoot2->left) && HasSubtree2(pRoot1->right, pRoot2->right);
    	}
	}
};